packages/alfa-rules/src/sia-er62/rule.ts

```diff
@@ -69,60 +61,128 @@ export default Rule.Atomic.of<Page, Element>({
+              containers = containers.set(node, container.get());
+              link = Option.of(node);
             }
 
             // Otherwise, if the element is a <p> element with non-link text
             // content then start collecting applicable elements.
-            else if (
-              test(
-                and(hasRole(device, "paragraph"), hasNonLinkText(device)),
-                node
-              )
-            ) {
+            if (isParagraph(node)) {
               container = Option.of(node);
             }
           }
 
+          const isTextNode = test(and(isText, isVisible(device)), node);
+          if (isTextNode && container.isSome()) {
+            // For each link, store the parent of the text nodes it contains
+            if (link.isSome()) {
+              linkText = linkText.set(
+                link.get(),
+                linkText
+                  .get(link.get())
+                  .getOr(Set.empty<Element>())
+                  .add(node.parent().filter(isElement).get())
+              );
+            }
+
+            // For each container, store the parent of the text nodes it contains
+            if (link.isNone()) {
+              nonLinkText = nonLinkText.set(
+                container.get(),
+                nonLinkText
+                  .get(container.get())
+                  .getOr(Set.empty<Element>())
+                  .add(node.parent().filter(isElement).get())
+              );
+            }
+          }
+
           const children = node.children({
             flattened: true,
             nested: true,
           });
 
           for (const child of children) {
-            yield* visit(child, container);
+            gather(child, container, link);
+          }
+        }
+
+        function* getApplicableLinks(): Iterable<Element> {
+          // Check if foreground is the same with the parent <p> element
+          const hasDifferentForeground = (
+            link: Element,
+            container: Element
+          ): boolean =>
+            getForeground(link, device).some((linkColors) => {
+              // If the link has a foreground with the alpha channel less than 1 and background gradient color
+              // then the rule is applicable as we can't tell for sure if it ever has the same foreground with a link
+              // that might have the same foreground and gradien background, but with different gradient type or spread
+              if (linkColors.length !== 1) {
+                return true;
+              }
+              return getForeground(container, device).some(
+                (containerColors) => {
+                  // Similalry to the link, we assume it's applicable if the container has different foreground colors
+                  if (containerColors.length !== 1) {
+                    return true;
+                  }
+                  return !linkColors[0].equals(containerColors[0]);
+                }
```

Comment:
Is it possible that in this case the second `getForeground` should be negated like this:
```Javascript
getForeground(link, device).none(
  (linkColors) =>
	linkColors.length === 1 &&
	!getForeground(container, device).none(
	  (containerColors) =>
		containerColors.length === 1 &&
		linkColors[0].equals(containerColors[0])
	)
);
```
In this case maybe this version is more readable:
```Javascript
getForeground(link, device).some(
  (linkColors) =>
	linkColors.length !== 1 ||
	getForeground(container, device).none(
	  (containerColors) =>
		containerColors.length === 1 &&
		linkColors[0].equals(containerColors[0])
	)
);
```

Reply:
> ```js
> getForeground(link, device).none(
>   (linkColors) =>
> 	linkColors.length === 1 &&
> 	getForeground(container, device).some(
> 	  (containerColors) =>
> 		containerColors.length === 1 &&
> 		linkColors[0].equals(containerColors[0])
> 	)
> );
> ```

This one should be good indeed.
The problem with the `foo.length !==1` tests is that they would pass if the length is 0 (not fully sure this does happen in practice), which would break the `foo[0]` later on.